### ISSUE TITLE
Iterate over dynamic slices

### ISSUE THREAD

**Issue by alewis** [google/jax#2308]
Hi there,

Excellent work here!

Our team is trying to construct a Jax implementation of a matrix factorization which requires us to iterate over slices whose dimensions depend upon the iteration counter. For example, suppose we need to operate upon each column of some matrix’s lower-diagonal, as in the Householder QR algorithm, along the lines of:

```Python
@jax.jit
def operate_on_lower_triangle(A):
    for j in A.size[1]:
        A_slice = A[j:, j]  # *
        v = vector_function(A_slice)
        A = jax.lax.index_update(A, index[j:, j], apply_transform(A, v))
```

This actually works, but since the for loop gets statically unrolled, compilation can get really slow, especially in the more complex example we are working with. The cure for this is to use one of the Lax control flow primitives. But this has the consequence that ‘j’ becomes an abstract tracer, so that the line marked `*` becomes problematic due to the dynamic shape of the slice.

Is there a workaround for this? It seems like there must be, since QR has a TPU implementation.

**Comment by hawkinsp** [comment]
Our current solution for this kind of compilation time problem is to at least partially roll up either inner or outer loops.

The QR implementation JAX uses is in the XLA client library, in C++:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/client/lib/qr.h

You can do the same sort of thing in Python too, e.g., the LU decomposition here: https://github.com/google/jax/blob/master/jax/lax_linalg.py#L452

You can use `lax.dynamic_update_slice` or `lax.scatter` if you need more control over updates; ultimately that's what the indexed update lowers to.

I think it's an open question how well these implementations can perform without more manual tuning, but they certainly suffice for many things.

Hope that helps!

**Comment by hawkinsp** [comment]
On CPU and GPU, you can also call into custom kernels, and indeed we do that for LAPACK and Cusolver kernels.

**Comment by hawkinsp** [comment]
Another trick that LU decomposition uses is masked updates to make some things that wouldn't have static shapes static once again.

**Comment by Thenerdstation** [comment]
Note: Adam is working with me on this.

**Comment by Thenerdstation** [comment]
I'm mainly curious as to why the compilation of QR if fast vs our python version is so slow. From our conversations in private, there really should not be a difference in compilation time between C++ XLA and python XLA from JAX. Am I misunderstanding?

**Comment by shoyer** [comment]
> I'm mainly curious as to why the compilation of QR if fast vs our python version is so slow. From our conversations in private, there really should not be a difference in compilation time between C++ XLA and python XLA from JAX. Am I misunderstanding?

That's right. This code would be just as slow to compile if you wrote a `for` loop for code generation in C++. The critical thing is really using XLA's control flow operations...

**Comment by alewis** [comment]
Thanks, this is helpful.

As you mentioned, it looks like in LU some of the slicing is handled using np.where to create masks. In these cases, do the masked entries contribute to the instruction count (e.g is A[k+1:]*2 using where cheaper than A[:]/2 would have been)?

**Comment by shoyer** [comment]
Yes, you still do the computation on the masked elements. But depending on the platform (particularly for accelerators like TPUs) this may even be faster than trying to do dynamic slicing, regardless of compilation times.

**Comment by mattjj** [comment]
I think this thread is stale, so I'll close it. Please correct me if I'm wrong!
